### PR TITLE
(feat) Patient lists UI fixes and improvements

### PR DESCRIPTION
**Pull request by denniskigen**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and [**design documentation**](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary

This PR applies various fixes and improvements to the Patient Lists UI. These include:

- Implementing fuzzy search in the DataTable search boxes.
- Reusing the `useDebounce` hook for debouncing searches.
- Adding various empty states for when a search returns no matching results.
- Renaming various components to better reflect their purpose.

See the relevant Zeplin screens [here](https://zpl.io/VOLvn8Y).

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-management/assets/8509731/3acd6c35-8164-44b6-9065-c10592e212e2

## Related Issue

*None*

## Other

*None*

### CODE REVIEW

**Review by ibacher — APPROVED** [review]
Some very minor suggestions, but overall LGTM!

File: packages/esm-patient-list-app/src/create-edit-patient-list/create-edit-list.component.tsx [hunk 1]
```diff
@@ -62,12 +60,10 @@ const CreateEditPatientList: React.FC<CreateEditPatientListProps> = ({
...
         .then(close)
-        .catch(() => {
+        .catch((e) => {
           showToast({
-            title: t('error', 'Error'),
-            description: `${t('errorCreatePatientListDescription', "Couldn't create patient list")} : ${
-              cohortDetails?.name
-            }`,
+            title: t('error', 'Error creating list'),
+            description: e?.message,
```
[the remark this answers is not in the document]
Comment (denniskigen) [reply packages/esm-patient-list-app/src/create-edit-patient-list/create-edit-list.component.tsx:66]: Might something like:

```tsx
  showToast({
    title: t('errorCreatingList', 'Unable to create list'),
    description: t('errorCreatingListDescription', 'The list could not be created. Please try again.'),
    kind: 'error',
  });
```

work? Inspired by the following example from the [notifications](https://zeroheight.com/23a080e38/p/683580-notifications) guide:

CleanShot 2023-11-01 at 8  50 32@2x

File: packages/esm-patient-list-app/src/header/header.component.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+const Header: React.FC = () => {
+  const { t } = useTranslation();
+  const newCohortUrl = window.getOpenmrsSpaBase() + 'home/patient-lists?new_cohort=true';
+
+  const handleShowNewListOverlay = () => {
+    // URL navigation is in place to know either to open the create list overlay or not
+    // The url /patient-list?new_cohort=true is being used in the "Add patient to list" widget
+    // in the patient chart. The button in the above mentioned widget "Create new list", navigates
+    // to /patient-list?new_cohort=true to open the overlay directly.
```
[the remark this answers is not in the document]
Comment (denniskigen) [reply packages/esm-patient-list-app/src/header/header.component.tsx:17]: Didn't make too much sense to me either. This was grabbed directly from https://github.com/openmrs/openmrs-esm-patient-management/blob/44fd35646288e072ad68dac25d6b5e62ce8ea9c7/packages/esm-patient-list-app/src/patient-list-list/patient-list-list.component.tsx#L65

[the remark this answers is not in the document]
Comment (ibacher) [reply packages/esm-patient-list-app/src/header/header.component.tsx:17]: @vasharma05 I think you wrote this... Any insight?

File: packages/esm-patient-list-app/src/create-edit-patient-list/create-edit-list.component.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibacher) [reply packages/esm-patient-list-app/src/create-edit-patient-list/create-edit-list.component.tsx:66]: It's fine. I'm actually ok with displaying the exception message to user. I just think we need some context so user's understand what the message is related to.

File: packages/esm-patient-list-app/src/api/api-remote.ts [hunk 3]
```diff
@@ -15,36 +15,42 @@ import {
 
 export const cohortUrl = '/ws/rest/v1/cohortm';
 
-async function postData(url: string, data = {}, ac = new AbortController()) {
```
Comment (ibacher) [inline packages/esm-patient-list-app/src/api/api-remote.ts:18]: I'd keep the `AbortController` parameter. `AbortController`s are used to signal to `fetch()` that the request can be cast aside because the component waiting for the data no longer needs it. The most obvious use-case for this is something like:

1. Component mounts and requests data
2. While data is being fetched, the component is unmounted due to user action

At 2, we want to stop the data request (or at least signal that its unimportant), but that means having an `AbortController` that's aware of the lifecycle of the containing component (which is kind of the point of the `useAbortController()` hook I added to the framework).

Basically, a purely local-variable `AbortController()` is just overhead, since it will never abort.

**Review by ibacher — APPROVED** [review]

File: packages/esm-patient-list-app/src/api/api-remote.ts (hunk 3 above)
Comment (denniskigen) [reply packages/esm-patient-list-app/src/api/api-remote.ts:18]: Yeah, that makes a lot of sense. I'll revert back to the previous behaviour.
